File: 13-eval-rpn/solution.cs
// Evaluate Reverse Polish Notation
// Time: O(n), Space: O(n)

using System;
using System.Collections.Generic;

public class Solution {
    public int EvalRPN(string[] tokens) {
        Stack<int> stack = new Stack<int>();
        HashSet<string> operators = new HashSet<string> { "+", "-", "*", "/" };
        
        foreach (string token in tokens) {
            if (operators.Contains(token)) {
                int b = stack.Pop();
                int a = stack.Pop();
                switch (token) {
                    case "+":
                        stack.Push(a + b);
                        break;
                    case "-":
                        stack.Push(a - b);
                        break;
                    case "*":
                        stack.Push(a * b);
                        break;
                    case "/":
                        stack.Push(a / b);
                        break;
                }
            } else {
                stack.Push(int.Parse(token));
            }
        }
        
        return stack.Pop();
    }
    
    public static void Main(string[] args) {
        Solution sol = new Solution();
        Console.WriteLine(sol.EvalRPN(new string[] {"2", "1", "+", "3", "*"}));  // 9
        Console.WriteLine(sol.EvalRPN(new string[] {"4", "13", "5", "/", "+"}));  // 6
    }
}

File: 13-eval-rpn/solution.java
// Evaluate Reverse Polish Notation
// Time: O(n), Space: O(n)

import java.util.*;

class Solution {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        Set<String> operators = new HashSet<>(Arrays.asList("+", "-", "*", "/"));
        
        for (String token : tokens) {
            if (operators.contains(token)) {
                int b = stack.pop();
                int a = stack.pop();
                switch (token) {
                    case "+":
                        stack.push(a + b);
                        break;
                    case "-":
                        stack.push(a - b);
                        break;
                    case "*":
                        stack.push(a * b);
                        break;
                    case "/":
                        stack.push(a / b);
                        break;
                }
            } else {
                stack.push(Integer.parseInt(token));
            }
        }
        
        return stack.pop();
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.evalRPN(new String[]{"2", "1", "+", "3", "*"}));  // 9
        System.out.println(sol.evalRPN(new String[]{"4", "13", "5", "/", "+"}));  // 6
    }
}
